### PR TITLE
Update to Angular 16, TypeScript 5, and latest Firebase SDK

### PR DESCRIPTION
**Pull request by davideast**
## What is this?
This PR aims to update AngularFire to full compatibility with Angular 16 and the latest Firebase features. However, this requires a jump to TypeScript 5 which requires a refactoring of our build system.

This will likely require a major version bump and as we will detail soon we are looking to match versions with Angular. Therefore the next AngularFire version could jump from 7 to 16. We'll have an open discussion in the near future but please let us know your thoughts.

### Addresses
#3320, #3347

### Todo
- [x] Update dependencies for Angular 16 and TypeScript 5
- [x] Update to ng-packagr 16
- [x] Test canary package
- [x] Zone wrap new Firebase SDK methods
- [ ] Refactor build system (In progress)
- [ ] Migrate from tslint to eslint (In progress)
- [ ] Resolve zone.js typing issues on build (In progress)

### CODE REVIEW

File: .github/workflows/test.yml
Comment (jhuleatt) [inline .github/workflows/test.yml:41]: maybe take these out if they're going to stay commented out?

Comment (jhuleatt) [reply .github/workflows/test.yml:41]: (same for all other restore-keys?)

File: angular.json
Comment (jhuleatt) [inline angular.json:46]: is it still true that the package manager is yarn?

File: src/schematics/deploy/actions.jasmine.ts
Comment (jhuleatt) [inline src/schematics/deploy/actions.jasmine.ts:4]: commented out line - remove?

**Review by jhuleatt — APPROVED** [review]
I'm not an Angular expert so mostly left nits. Great that things are green, though!

File: tsconfig.json
Comment (davideast) [reply tsconfig.json:11]: This is temporary as I ran into typing issues with Zone that only error out on CI but not locally. I'm working on a local GitHub Actions setup that will make fixing issues like these easier in the future.

File: firebase.json
Comment (davideast) [reply firebase.json:23]: Yep! Just needed something less common.

File: angular.json
Comment (davideast) [reply angular.json:46]: Yep! We could migrate off but no need to right now.

File: .github/workflows/test.yml
Comment (davideast) [reply .github/workflows/test.yml:41]: Good call. I'll remove all of this for now and work rebuilding the matrix in the future.

**Comment by yharaskrik** [comment]
I am very excited for this PR. Like REALLY excited.

**Comment by akikesulahti** [comment]
Version jump from 7 to 16 sounds good. Material and CLI are already matching versions with Angular and I really like that approach - eg. makes compatibility checking easier. And the concept is clear so I don't see any drawbacks if Angularfire will also match version with Angular.

**Comment by adelamata** [comment]
Good job, I wait it with a big smiling!!
Thanks!!

**Comment by davideast** [comment]
Hey everyone,

Sorry to have held this up. I've had some other tasks come to the front but I'll be back getting this across the line. I need to fix a bug in RxFire and then upgrade the library to that version so everyone can have COUNT(*) as an observable.

**Comment by trakhimenok** [comment]
> so everyone can have COUNT(*) as an observable.

Maybe you can release version without `COUNT(*)` for those who does not use it?

**Comment by jits** [comment]
Hello. Is there anything we can do to help speed this up?

**Comment by tgangso** [comment]
When can we expect this? There is an important update in firebase 10.3.0 that is dependent on this case. https://github.com/firebase/firebase-js-sdk/issues/6118

**Comment by davideast** [comment]
### tldr;
 A bump to 16.0.0. This will merge and release as soon as we cut a new RxFire version on August 28th.

### What's going on?
Alright! This release is almost ready to go. The only blocker is getting an RxFire release out that will fix some bugs upstream here in AngularFire. The PR currently is tested with a canary of RxFire (all green!) and once RxFire is released we'll test again and merge. 

### What didn't make it?
The following items below aren't going to make the release because they are infrastructure for the library and not directly developer impacting. I don't want to block the release on them as they are tedious and time consuming tasks that have are mostly covered by unit tests downstream. We'll get to this in an upcoming release and it shouldn't require a version bump. 

- Refactor build system
- Migrate from tslint to eslint
- Resolve zone.js typing issues on build

**Comment by yharaskrik** [comment]
> ### tldr;
> A bump to 16.0.0. This will merge and release as soon as we cut a new RxFire version on August 28th.
> 
> ### What's going on?
> Alright! This release is almost ready to go. The only blocker is getting an RxFire release out that will fix some bugs upstream here in AngularFire. The PR currently is tested with a canary of RxFire (all green!) and once RxFire is released we'll test again and merge.
> 
> ### What didn't make it?
> The following items below aren't going to make the release because they are infrastructure for the library and not directly developer impacting. I don't want to block the release on them as they are tedious and time consuming tasks that have are mostly covered by unit tests downstream. We'll get to this in an upcoming release and it shouldn't require a version bump.
> 
> * Refactor build system
> * Migrate from tslint to eslint
> * Resolve zone.js typing issues on build

Amazing! Thank you David, everyone in this thread really really appreciates it!

**Comment by davideast** [comment]
It is done! Can someone use the canary build and let me know how it's working for them?

```
npm i @angular/fire@canary
```

**Comment by yharaskrik** [comment]
> It is done! Can someone use the canary build and let me know how it's working for them?
> 
> ```
> npm i @angular/fire@canary
> ```

On it.

**Comment by yharaskrik** [comment]
> It is done! Can someone use the canary build and let me know how it's working for them?
> 
> ```
> npm i @angular/fire@canary
> ```

Haven't checked every app of ours yet but it seems to be working! builds pass, app loads with authentication working. Tried with the webpack and esbuild builder.

**Comment by atsjo** [comment]
doesn't have the correct firebase dependency from what I can see... rxfire has ^10.0.0 as peer and @angular/fire has ^9.8.0 as dep, giving a warning for incorrect peer dependency...

**Comment by ginagr** [comment]
I just updated to `@angular/fire@canary` and get the same error as before. I did a clean install just to make sure it wasn't a cache issue but still unfortunately see the same error. 
Screenshot 2023-08-28 at 3 59 23 PM

**Comment by atsjo** [comment]
It doesn't seem to zonewrap getCountFromServer firestore function from what I can see either, not sure if zonewrapping new functions should be part of this release... I don't use the rxfire wrappers, I just want the zonewrapped functions from the firebase sdk...

**Comment by davideast** [comment]
> I just updated to `@angular/fire@canary` and get the same error as before. I did a clean install just to make sure it wasn't a cache issue but still unfortunately see the same error. Screenshot 2023-08-28 at 3 59 23 PM

Do you have a way for me to reproduce this?

**Comment by davideast** [comment]
> It doesn't seem to zonewrap getCountFromServer firestore function from what I can see either, not sure if zonewrapping new functions should be part of this release... I don't use the rxfire wrappers, I just want the zonewrapped functions from the firebase sdk...

My mistake. I only added the [observable wrappers](https://github.com/angular/angularfire/blob/master/src/firestore/rxfire.ts#L26-L27) on this release but I can also add `getCountFromServer()` as well.

File: src/schematics/ng-add.jasmine.ts
Comment (robertIsaac) [inline src/schematics/ng-add.jasmine.ts:2]: you forget this comment

File: tools/run-typings-test.js
Comment (robertIsaac) [inline tools/run-typings-test.js:51]: shouldn't it be `console.error`

File: tsconfig.build.json
Comment (robertIsaac) [inline tsconfig.build.json:9]: why commonjs?

File: src/compat/auth/auth.spec.ts
Comment (robertIsaac) [inline src/compat/auth/auth.spec.ts:2]: `forkJoin, merge, tap` aren't used, should be removed

**Comment by robertIsaac** [comment]
> ### tldr;
> A bump to 16.0.0. This will merge and release as soon as we cut a new RxFire version on August 28th.
> 
> ### What's going on?
> Alright! This release is almost ready to go. The only blocker is getting an RxFire release out that will fix some bugs upstream here in AngularFire. The PR currently is tested with a canary of RxFire (all green!) and once RxFire is released we'll test again and merge.
> 
> ### What didn't make it?
> The following items below aren't going to make the release because they are infrastructure for the library and not directly developer impacting. I don't want to block the release on them as they are tedious and time consuming tasks that have are mostly covered by unit tests downstream. We'll get to this in an upcoming release and it shouldn't require a version bump.
> 
> * Refactor build system
> * Migrate from tslint to eslint
> * Resolve zone.js typing issues on build

I can work on eslint, I have experience on it, but won't work on it unless I have green light so I don't waste time on it while it won't be merged
I'd use eslint flat configuration with eslint/js recommended and typescript recommended requiring type checking and import sort to make sure all import are sorted, will turn off any rule that will cause a lot of changes, and can be addressed in a future PR as standalone

**Comment by tgangso** [comment]
I get an error when trying to use:

firebase.firestore.FieldValue.serverTimestamp();

TypeError: Cannot read properties of undefined (reading 'FieldValue')

and

firebase.analytics().logEvent

Error logEvent firebase TypeError: firebase_compat_app__WEBPACK_IMPORTED_MODULE_1__.default.analytics is not a function


Errors goes away if I downgrade to v9.

**Comment by davideast** [comment]
> I can work on eslint, I have experience on it, but won't work on it unless I have green light so I don't waste time on it while it won't be merged I'd use eslint flat configuration with `eslint/js/recommended` and `typescript/recommended-requiring-type-checking`, `angular-eslint` and `import` to make sure all import are sorted, will turn off any rule that will cause a lot of changes, and can be addressed in a future PR as standalone

This would be amazing! If you get a PR going I'll get it merged.

**Comment by robertIsaac** [comment]
> It is done! Can someone use the canary build and let me know how it's working for them?
> 
> ```
> npm i @angular/fire@canary
> ```

works perfectly for me

**Comment by jits** [comment]
Testing the `16.0.0-canary.596e208` version on a small prototype (using Yarn v1)…

- `yarn add @angular/fire@canary`
- Then had to manually bump `rxfire` in `yarn.lock` to `6.0.4` as it was stuck on `6.0.3` (by deleting the whole rxfire section and running `yarn install` again).
- Then had to force the resolution of `firebase` to latest v10 (in `package.json`).
- Then the following all worked (communicating to local emulator only):
  - ✅ Auth log in and log out (username/password only)
  - ✅ RTDB get
  - ✅ Firestore CRUD
  - ✅ Callable function

**Comment by ginagr** [comment]
> > I just updated to `@angular/fire@canary` and get the same error as before. I did a clean install just to make sure it wasn't a cache issue but still unfortunately see the same error. Screenshot 2023-08-28 at 3 59 23 PM
> 
> Do you have a way for me to reproduce this?

I tried making a minimal preproduction but everything seemed to run fine, so maybe it's just a me issue ¯\ _(ツ)_/¯
Thanks for all the hard work you've put into this!

**Comment by JakobMadsen717** [comment]
I npm install these:
@angular/fire 16.0.0-canary.e04cd7f
firebase 10.3.1
in my angular (16.2.3) app. 

I get these compilation errors:

Error: src/app/services/user.service.ts:30:29 - error TS2345: Argument of type 'import(".../node_modules/@angular/fire/auth/auth").Auth' is not assignable to parameter of type 'import(".../node_modules/@firebase/auth/dist/auth-public").Auth'.
  Property 'authStateReady' is missing in type 'import(".../node_modules/@angular/fire/auth/auth").Auth' but required in type 'import(".../node_modules/@firebase/auth/dist/auth-public").Auth'.

    private auth: Auth
    ..
    this.fbUsr$ = authState(this.auth).pipe(
                            ~~~~~~~~~

  node_modules/@firebase/auth/dist/auth-public.d.ts:397:5
    397     authStateReady(): Promise<void>;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    'authStateReady' is declared here.

The odd thing is: if roll back firebase to version 10.0.0 instead of 10.3.1 then compilation errors disappear and app works.

I don't have a repo to share unfortunately. It could be a code problem of course, but I wanted to let you know in case it means anything to your efforts. Thanks for working on this upgrade.

**Comment by kanehjeong** [comment]
I have the same issue as @JakobMadsen717 
Screenshot 2023-09-01 190536

**Comment by atsjo** [comment]
The version range change had no effect on the build output... I tested 16.0.0-canary.5793d6f, and it seems to still have
"firebase": "^9.8.0",
"rxfire": "^6.0.0",
as dependencies, causing error when building with firebase 10.x. I resolve this via overrides in pnpm....

**Comment by atsjo** [comment]
changing the firebase dep in https://github.com/angular/angularfire/blob/master/src/package.json to
"firebase": "^9.8.0 || ^10.0.0"
should fix this...

**Comment by anisabboud** [comment]
`ng add @angular/fire@canary` produces an error in schematics:
```Collection ".../node_modules/@angular/fire/schematics/collection.json" cannot be resolved.```

**Comment by anisabboud** [comment]
rxfire 6.0.5 produces typing error:
```bash
Error: node_modules/rxfire/firestore/lite/interfaces.d.ts:8:29 - error TS2314: Generic type 'AggregateQuerySnapshot<T>' requires 1 type argument(s).

 8 export type CountSnapshot = lite.AggregateQuerySnapshot<{
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 9     count: lite.AggregateField<number>;
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10 }, any, DocumentData>;
   ~~~~~~~~~~~~~~~~~~~~~
```

**Comment by rgant** [comment]
@anisabboud This is how I was able to resolve that issue:
https://github.com/FirebaseExtended/rxfire/issues/88#issuecomment-1715851191

**Comment by anisabboud** [comment]
👀

**Comment by tgangso** [comment]
16.0.0-rc.0 seems to be working fine here and fixed the issues I mentioned.

**Pull request merged**
